File: tests/python/src/publisher.py
import sys
import cameopy

this = cameopy.This
this.init(sys.argv)

numberOfSubscribers = 1 if len(sys.argv) < 3 else int(sys.argv[1])
print("Number of subscribers", numberOfSubscribers)

print("Creating publisher and waiting for ", numberOfSubscribers, "subscriber(s)...")

publisher = cameopy.coms.Publisher.create("publisher")
publisher.setWaitForSubscribers(numberOfSubscribers)
publisher.init()

this.setRunning()

print("Synchronized with ", numberOfSubscribers, "subscriber(s)")

for i in range(100):
    message = "message " + str(i)  
    publisher.send(message)

publisher.sendEnd()

publisher.terminate()
this.terminate()

print("Finished the application")

File: tests/cpp/src/Publisher.cpp
/*
 * CAMEO
 *
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under BSD 3-Clause and GPL-v3 as described in license files.
 * You may not use this work except in compliance with the Licences.
 *
 */

#include <cameo/api/cameo.h>
#include <iostream>
#include <sstream>

using namespace std;
using namespace cameo;

int main(int argc, char *argv[]) {

	This::init(argc, argv);

	bool syncSubscribers = false;
	int numberOfSubscribers = 1;

	if (argc > 2) {
		syncSubscribers = (string(argv[1]) == "true");
	}

	if (argc > 3) {
		istringstream is(argv[2]);
		is >> numberOfSubscribers;
	}

	cout << "Number of subscribers is " << numberOfSubscribers << endl;
	cout << "Sync subscribers is " << syncSubscribers << endl;

	unique_ptr<coms::Publisher> publisher;

	try {
		cout << "Creating publisher and waiting for " << numberOfSubscribers << " subscriber(s)..." << endl;

		publisher = coms::Publisher::create("publisher");
		publisher->setSyncSubscribers(syncSubscribers);
		publisher->setWaitForSubscribers(numberOfSubscribers);
		publisher->init();
	}
	catch (const InitException& e) {
		cout << "Publisher error" << endl;
		return -1;
	}

	This::setRunning();

	cout << "Synchronized with " << numberOfSubscribers << " subscriber(s)" << endl;

	// Sending data.
	for (int i = 0; i < 100; ++i) {
		publisher->send("message " + to_string(i));
	}

	publisher->sendEnd();

	cout << "Finished the application" << endl;

	return 0;
}